buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy

```diff
@@ -135,49 +144,46 @@ class BuildPlugin implements Plugin<Project> {
-                        } else {
-                            update = matcher.group(1).toInteger()
-                        }
-                        if (update < 40) {
-                            throw new GradleException("JDK ${javaVendor} ${javaVersion} has compiler bug JDK-8052388, update your JDK to at least 8u40")
-                        }
-                    }
-                }
+            if (runtimeJavaVersionEnum < minimumRuntimeVersion) {
+                throw new GradleException("Java ${minimumRuntimeVersion} or above is required to run Elasticsearch")
             }
 
-            project.rootProject.ext.javaHome = javaHome
-            project.rootProject.ext.javaVersion = javaVersionEnum
+            project.rootProject.ext.compilerJavaHome = compilerJavaHome
+            project.rootProject.ext.runtimeJavaHome = runtimeJavaHome
+            project.rootProject.ext.compilerJavaVersion = compilerJavaVersionEnum
+            project.rootProject.ext.runtimeJavaVersion = runtimeJavaVersionEnum
             project.rootProject.ext.buildChecksDone = true
         }
-        project.targetCompatibility = minimumJava
-        project.sourceCompatibility = minimumJava
+        project.targetCompatibility = minimumRuntimeVersion
+        project.sourceCompatibility = minimumRuntimeVersion
         // set java home for each project, so they dont have to find it in the root project
-        project.ext.javaHome = project.rootProject.ext.javaHome
-        project.ext.javaVersion = project.rootProject.ext.javaVersion
+        project.ext.compilerJavaHome = project.rootProject.ext.compilerJavaHome
+        project.ext.runtimeJavaHome = project.rootProject.ext.runtimeJavaHome
+        project.ext.compilerJavaVersion = project.rootProject.ext.compilerJavaVersion
+        project.ext.runtimeJavaVersion = project.rootProject.ext.runtimeJavaVersion
     }
 
     /** Finds and enforces JAVA_HOME is set */
-    private static String findJavaHome() {
-        String javaHome = System.getenv('JAVA_HOME')
+    private static String findCompilerJavaHome() {
+        return findJavaHome(System.getenv('JAVA_HOME'), null)
+    }
+
+    private static String findRuntimeJavaHome(final String compilerJavaHome) {
+        return findJavaHome(System.getenv('RUNTIME_JAVA_HOME'), compilerJavaHome)
+    }
+
+    private static String findJavaHome(String maybeJavaHome, String defaultJavaHome) {
+        final String javaHome = maybeJavaHome ?: defaultJavaHome
         if (javaHome == null) {
             if (System.getProperty("idea.active") != null || System.getProperty("eclipse.launcher") != null) {
-                // intellij doesn't set JAVA_HOME, so we use the jdk gradle was run with
+                // IntelliJ does not set JAVA_HOME, so we use the JDK that Gradle was run with
                 javaHome = Jvm.current().javaHome
             } else {
-                throw new GradleException('JAVA_HOME must be set to build Elasticsearch')
```

Comment:
Why was this exception changed to an assertion? Is it moved somewhere else I'm missing?

Reply:
I am glad that you caught this. This was inadvertent, it arose in a refactoring I did in 95475f922ba56017dc93fde0945cafe6884c44b2 which refactored two basically identical methods from an earlier version of this pull request (`findCompilerJavaHome` and `findRuntimeJavaHome`) into one (`findJavaHome`). Before the refactoring, the second method did not need such a line because by the time it had executed the first one would have already thrown that exception (hence the `assert false`). When I refactored, I must have based by work on `findRuntimeJavaHome` inadvertently losing this exception. I have pushed: bcee920a74dd3b9a9e6ae6d39655ecaea7ca486c